feat(common):focus directive

AlmogShaul opened the pull request.

AlmogShaul committed:
focus directive

AlmogShaul committed:
focus directive

AlmogShaul committed:
copyrights and removed duplicate imports

AlmogShaul committed:
reorders imports

AlmogShaul committed:
reorders imports

zoechi commented on modules/@angular/common/src/directives/ng_focus.ts:
```diff
@@ -0,0 +1,48 @@
+
+  protected ngOnChanges() {
+    if (this.focus)
+      this.element.nativeElement.focus();
```
Accessing members of `nativeElement` directly is bad practice. See http://stackoverflow.com/questions/34522306/angular-2-focus-on-newly-added-input-element for how this can be done without calling `blur()` and `focus()` on `nativeElement`

DzmitryShylovich replied:
Hmm https://github.com/angular/material2/blob/master/src/lib/input/input.ts#L196 :)

zoechi replied:
@DzmitryShylovich I also wondered about this when I saw it a while back ...

AlmogShaul committed:
using Renderer

AlmogShaul committed:
using Renderer

AlmogShaul committed:
clang-formatting

AlmogShaul committed:
added export to the TestComponent

AlmogShaul committed:
added export to the TestComponent

AlmogShaul committed:
add function main

AlmogShaul committed:
add function main

AlmogShaul commented:
@zoechi  Can you help figure out why 
I am getting errors in the Travis CI?

zoechi commented:
@AlmogShaul sorry, no. I don't know about the automated build and testing in TS.
Maybe someone else.

sirajc commented:
I think you should use `TestBed` instead of `TestComponentBuilder`. I saw lot of commits from core team moving from TCB to TestBed

AlmogShaul committed:
add function main

AlmogShaul committed:
add function main

AlmogShaul committed:
Merge branch 'master' of https://github.com/angular/angular into directives

AlmogShaul committed:
changed to TestBed

AlmogShaul committed:
add function main

AlmogShaul committed:
add function main

AlmogShaul committed:
add function main

AlmogShaul committed:
add function main

AlmogShaul committed:
add function main

AlmogShaul committed:
add function main

AlmogShaul committed:
add function main

AlmogShaul committed:
check other failures

AlmogShaul committed:
check other failures

AlmogShaul commented:
@sirajc @zoechi Can you please have a look at my code, I cant understand what is the problem
I changed to TestBed, added NgModule and it just fails in the Travis

jmcgoldrick commented:
Hey @AlmogShaul, RE the build failures, check out the logs on Travis:

`Can't bind to 'ngFocus' since it isn't a known property of ...` and work backwards from there, but:

If you think `NgFocus` shoud be part of `@angular/common` I would suggest writing the tests just like the other common directives ([example](https://github.com/angular/angular/blob/master/modules/%40angular/common/test/directives/ng_style_spec.ts)) - note how `CommonModule` (which declares/exports the directive) is imported and **not** the individual directive under test - something like this also ensures the inclusion in the `NgModule` is valid. You might just find this resolves your error(s) too. :smile:

You'll need to clean up all the style changes you have made in `public_api_guard/common/index.d.ts` too once the tests are passing.

I also wonder if something like this belongs in `@angular/common` or perhaps could/should be available in a third-party module - how would this interact with something like the focus-trapping requirements etc. in the angular2-material project for example? That being said, I have no idea what the Angular teams' intentions for their modules are...

AlmogShaul commented:
Thanks.I will check it.
BTW who decide which directive get in to the common?

mhevery commented on modules/@angular/common/src/directives/ng_focus.ts:
```diff
@@ -0,0 +1,49 @@
+@Directive({selector: '[ngFocus]'})
+export class NgFocus {
+  @Input()
+  focus: boolean;
```
Turn this into a setter, it is more efficient then `ngOnChanges`

`set focus(value: any) { ... }`

mhevery commented on tools/public_api_guard/common/index.d.ts:
```diff
@@ -1,233 +1,230 @@
+import {Renderer} from '../../../modules/@angular/core/src/render/api';
```
this file should not be changed and needs to be reverted.

mhevery commented:
Mechanical things:
1. Squash all SHA's into one
2. revert the public API file
3. get tests to pass. They are failing because you need to add `NgFocus` to `core_directives.ts`

Philosophical Things:
1. So far everything in `core_directives` is HTML agnostic and we are on the fence wether we should include DOM specific things such as focus. If we do, then maybe renderer should be focus aware, just like renderer is `class` and `style` aware. This has to be thought through a bit more. 

2. The other thing is that a focus is an action not a binding, for that reason it is weird that we use change detection to trigger it. Again more thought needed as the direction we should take.

AlmogShaul committed:
check other failures

AlmogShaul committed:
Merge branch 'master' of https://github.com/angular/angular into directives

# Conflicts:
#	tools/public_api_guard/common/index.d.ts

AlmogShaul committed:
Merge branch 'master' of https://github.com/angular/angular into directives

# Conflicts:
#	tools/public_api_guard/common/index.d.ts

AlmogShaul committed:
Merge branch 'master' of https://github.com/angular/angular into directives

# Conflicts:
#	tools/public_api_guard/common/index.d.ts

AlmogShaul commented:
@mhevery
I get errors on Travis:
I's very strange, I reverted all my changed in the public_api_guard/common/index.d.ts file,
how can it be that I am getting errors?


+node dist/tools/@angular/tsc-wrapped/src/main -p modules
Error: Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:2:37: Cannot find name 'OpaqueToken'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:5:43: Cannot find name 'OnDestroy'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:6:23: Cannot find name 'ChangeDetectorRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:8:20: Cannot find name 'Observable'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:8:53: Cannot find name 'EventEmitter'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:16:46: Cannot find name 'PipeTransform'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:22:42: Cannot find name 'PipeTransform'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:28:45: Cannot find name 'PipeTransform'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:47:48: Cannot find name 'PipeTransform'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:55:48: Cannot find name 'PipeTransform'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:62:42: Cannot find name 'PipeTransform'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:106:47: Cannot find name 'PipeTransform'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:111:41: Cannot find name 'DoCheck'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:116:35: Cannot find name 'IterableDiffers'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:116:70: Cannot find name 'KeyValueDiffers'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:116:94: Cannot find name 'ElementRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:116:117: Cannot find name 'Renderer'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:121:39: Cannot find name 'DoCheck'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:121:48: Cannot find name 'OnChanges'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:123:20: Cannot find name 'TemplateRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:124:19: Cannot find name 'TrackByFn'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:125:33: Cannot find name 'ViewContainerRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:125:65: Cannot find name 'TemplateRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:125:106: Cannot find name 'IterableDiffers'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:125:129: Cannot find name 'ChangeDetectorRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:127:26: Cannot find name 'SimpleChanges'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:133:33: Cannot find name 'ViewContainerRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:133:65: Cannot find name 'TemplateRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:145:40: Cannot find name 'SwitchView'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:151:42: Cannot find name 'TemplateRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:151:78: Cannot find name 'ViewContainerRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:155:41: Cannot find name 'DoCheck'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:159:27: Cannot find name 'KeyValueDiffers'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:159:51: Cannot find name 'ElementRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:159:74: Cannot find name 'Renderer'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:171:32: Cannot find name 'ViewContainerRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:171:63: Cannot find name 'TemplateRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:176:32: Cannot find name 'ViewContainerRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:176:63: Cannot find name 'TemplateRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:180:50: Cannot find name 'OnChanges'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:182:23: Cannot find name 'TemplateRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:183:36: Cannot find name 'ViewContainerRef'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:201:45: Cannot find name 'PipeTransform'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:221:43: Cannot find name 'PipeTransform'.
Error at /home/travis/build/angular/angular/tools/public_api_guard/common/index.d.ts:226:47: Cannot find name 'PipeTransform'.
    at check (/home/travis/build/angular/angular/dist/tools/@angular/tsc-wrapped/src/tsc.js:30:15)
    at Tsc.typeCheck (/home/travis/build/angular/angular/dist/tools/@angular/tsc-wrapped/src/tsc.js:76:9)
    at /home/travis/build/angular/angular/dist/tools/@angular/tsc-wrapped/src/main.js:29:23
    at process._tickCallback (node.js:401:9)
    at Function.Module.runMain (module.js:431:11)
    at startup (node.js:139:18)
    at node.js:999:3
Compilation failed

AlmogShaul committed:
Fixed imports

AlmogShaul commented:
@mhevery
I added my directive to the core_directives.ts, added reference to the CommonModule
but I still fail on Travis .

AlmogShaul committed:
Fixed imports

AlmogShaul committed:
Fixed imports

AlmogShaul committed:
Fixed imports

AlmogShaul committed:
Fixed imports

AlmogShaul committed:
Fixed imports

AlmogShaul committed:
Fixed imports

AlmogShaul committed:
Fixed imports

AlmogShaul committed:
Fixed imports

AlmogShaul commented:
@shlomiassaf Hi man, I saw you have PR with a new directive and you passed all tests.
Can you help me please with mine?
I keep failing on the tests, I get :
 Failed: Template parse errors:
    Can't bind to 'ngFocus' since it isn't a known property of 'div'. ("<div [ERROR ->][ngFocus]="inFocus" ></div>"): TestComponent@0:5

thanks man

AlmogShaul committed:
Revert: Fixed imports

AlmogShaul committed:
Merge branch 'master' of https://github.com/angular/angular into directives

AlmogShaul committed:
Revert: Fixed imports

AlmogShaul committed:
focus directive

AlmogShaul committed:
focus directive

AlmogShaul committed:
Merge remote-tracking branch 'origin/directives' into directives

# Conflicts:
#	modules/@angular/common/src/directives/ng_focus.ts
#	modules/@angular/common/test/directives/ng_focus_spec.ts

AlmogShaul committed:
focus directive

AlmogShaul committed:
Merge branch 'focus' into directives

AlmogShaul committed:
focus directive

AlmogShaul committed:
focus directive

AlmogShaul committed:
focus directive

AlmogShaul committed:
focus directive

AlmogShaul committed:
Revert: focus directive

AlmogShaul committed:
Merge branches 'directives' and 'focus' of https://github.com/AlmogShaul/angular into directives

AlmogShaul committed:
Revert: focus directive

asnowwolf commented:
I think it's not a good idea to ship it as a built-in directive. could you please give me a "user story" to help me understand it?

AlmogShaul commented:
Lets say you have a form that you want to enable in a specific case. And you want to focus a specific input there. The focus directive will be very useful.
BYW I am struggling with the tests of this directive. I keep failing there saying that he cant recognize the directive.
I added the directive to the core_directives and to the directives files
What I am missing?

AlmogShaul commented:
@asnowwolf 
Another user story : Lets say you have a component that hold inside it several inputs with ngSwitch between them. I want that when the user will focus the component , the focus will tunneled  into the inside input, if I won't take care of it, the user will have to click twice to get into the input

AlmogShaul committed:
Revert: focus directive

AlmogShaul committed:
Revert: focus directive

AlmogShaul committed:
Revert: focus directive

asnowwolf commented:
I meant that: It's not necessary to ship with angular-core. almost of directives in ng 1, such as ng-click/ng-mouseover/ng-focus, is replaced with data binding.

My question is: what's the differences between ng-focus and ng-click? why it should ship with the core?

AlmogShaul commented:
ng-click is binding the click event to angular as far as I know.
ngFocus is binding a component property to the element focus , there is a big difference.

asnowwolf commented:
Oh, I see. So, It should be SetFocus in fact, right? 

Even though, I think it shouldn't ship with the core yet, it should be a great third-party directive.

AlmogShaul commented:
ok thanks :)
BTW Is there a place for such 3rd parties?
